OHHTTPStubs/UnitTests/Test Suites/NSURLSessionTests.m

```diff
@@ -116,13 +119,15 @@ - (void)_test_redirect_NSURLSession:(NSURLSession*)session
             return [[[request URL] path] isEqualToString:@""];
         } withStubResponse:^OHHTTPStubsResponse *(NSURLRequest *request) {
             NSDictionary *headers = @{ @"Location": @"foo://unknownhost:666/elsewhere" };
-            return [[OHHTTPStubsResponse responseWithData:[[NSData alloc] init] statusCode:301 headers:headers]
+            NSLog(@"Stop here");
+            return [[OHHTTPStubsResponse responseWithData:[[NSData alloc] init] statusCode:httpStatusCode headers:headers]
```

Comment:
Maybe here instead of returning an empty `NSData`, we could use a data for the redirect that is computed from the original `httpBody` (maybe not exactly the _same_ data to be sure not to confuse them both, but maybe one derived from it) — And for those tests, don't use the `json` parameter then.

So make those stub be so that:

* for the original request (the stub on that line 123), we send a redirect response (with `httpStatusCode`, being 301, 302, …) with the body derived from the request: `stubbedRedirectResponse.httpBody = f(originalRequest.httpBody)`.
* Then the second stub handling the redirect (below, line ~129), we make it so the stubbed final response has a body re-derived from the redirect body: `finalResponse.httpBody = g(redirectResponse.httpBody) = g(f(originalRequest.httpBody))`

That way, once in the `completion` block at the call site, we can look at the resulting body and do some assertions there.

For example `f` can be so it returns a simple JSON representation of a dictionary like `{ "originalBody": originalBodyHere, "originalMethod": originalMethodHere }`. Then `g` can be so it returns a JSON wrapping that first JSON, like `{ "redirectedBody": { "originalBody": originalBodyHere, "originalMethod": originalMethodHere }, "redirectedMethod": redirectedMethodHere }`.
That would make the assertions in the actual tests below easy to ensure we didn't lose anything in the way.

Reply:
If what you're saying above in reference to below ~129 is true, then I expect that we'd be hitting the return on the withStubResponse. I observe that we're not hitting that in these tests. Am I confused when you refer to "...second stub handling the redirect (below, line ~129)"?
```
        [OHHTTPStubs stubRequestsPassingTest:^BOOL(NSURLRequest *request) {
            return [[[request URL] path] isEqualToString:@"/elsewhere"];
        } withStubResponse:^OHHTTPStubsResponse *(NSURLRequest *request) {
            return [[OHHTTPStubsResponse responseWithJSONObject:blockData statusCode:200 headers:nil]
                    requestTime:kRequestTime responseTime:kResponseTime];
        }];
```

Reply:
I had a comment here, but it became a bit irrelevant after taking a break. I've been looking at this code for too long now. ;)

Reply:
Haha too late, I saw it in my email notifications 😂 
(Thanks for telling me you deleted it though, as I would probably have gone crazy trying to find it on GitHub after receiving the email notification otherwise 😄)

Reply:
Indeed I think so too

Reply:
ok, we have two "withStubResponse" methods here. One for the original redirect, one for the second response (from the redirected request) for the path, @"/elsewhere". In the first redirected response, I'm trying to hijack the data payload with my own JSON stuff to keep track of originating request stuff (I think this is what you were wanting me to do). Then put that hijacked data payload on the wire for the redirected request, and hence, when we finally get the 200 response, we should have our built up JSON. However, it seems that apple is ignoring my change in data. I'm getting the original data posted all the way through. This isn't bad behavior and it's preventing man-in-the-middle attacks. It seems I won't be able tweak the HTTPBody in the response. 
Thoughts?

Reply:
I was actually thinking about it the other day. The way we implemented it is that when there is a redirect, we do redirect with the same body as the original. That's the whole point of your PR. That's what you implemented. When there's a 301, we redirect to the new URL but with the original request's body. So that's our implementation after all.

So it's logical, from that implementation we did, that, whatever body we return in the redirection response, our implementation redirects to the new URL with the original body after all 😉 

More than that, I think in our reasoning for those tests we confused the request and the response. In the real world, when you post a request to the server with some http body B1, if the server returns a 301 response, that 301 response doesn't have any http body itself. It's the server's redirect response after all

So in our specific context:

* the first stub should just return an `OHHTTPStubsResponse` with status code 301 (and no http body whatsoever, that body would be ignored anyway, and we respect the RFC on that)
* then the second stub is supposed to be the final response that the server would return, when queried on the redirected URL with that original body. So there we could make our fake response to be some JSON wrapping the original body in that case (like if we had a real server whose only role was to echo what we send to it, just wrapped inside a JSON)

Reply:
For the other response codes as well ...should we return null for the body in the first stub or only in 301?

Reply:
I think all redirect responses are supposed to have an empty body